src/Microsoft.Extensions.Configuration/MemoryConfigurationSource.cs

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.Configuration.Memory
+{
+    public class MemoryConfigurationSource : IConfigurationSource
+    {
+        public IEnumerable<KeyValuePair<string, string>> InitialData { get; set; }
+
+        public IConfigurationProvider Build(IConfigurationBuilder builder)
+        {
+            return new MemoryConfigurationProvider(InitialData);
+        }
+
+        private class MemoryConfigurationProvider :
+            ConfigurationProvider,
+            IEnumerable<KeyValuePair<string, string>>
```

Comment:
The `IEnumerable<KeyValuePair<string, string>>` and `Add(key,value)` are relocated to the MemoryConfigurationSource